### DISCUSSION TITLE
Rules not running when AP (wifi) is disconnected

### DISCUSSION THREAD

**Discussion by SimonFili** [arendst/Tasmota#19719] [Q&A]
Hi all,

I have an ESP32, controlling 2 appliances.  I have 2 rules for turning on/off those appliances according to a schedule.

Everything is working fine when the Wi-Fi is connected.
But If I disconnect the Wi-Fi, the rules simply do not run.

I was under the impression that they would work without Wi-Fi.

Am I wrong?

**Comment by sfromis** [comment]
"According to a schedule" needs Tasmota to have time of day, normally via a NTP server accessed via Wifi.

**Reply by barbudor** [reply]
May be sharing your rules would make it clear what is "according to a schedule"

**Comment by SimonFili** [comment]
Here are the rules:

```
Rule1:
rule1 ON Time#Minute=360 DO power1 on ENDON ON Time#Minute=1080 DO power1 off ENDON

Rule2:
rule2 on Time#Minute|5 do power2 on endon

And a Pulsetime:
pulsetime2 130
```

I was wondering if Wificonfig which is defaulted to 4, could be set to something else.
I don't have a physical console to check if Tasmota is in a loop trying to connect to Wi-Fi and this is why the rules do not start?

I can expect the NTP time to drift off if not connected a long time, but normally, the internal clock would run without Wi-Fi ?

Thanks!

**Reply by Jason2866** [reply]
Tasmota does not work without a time set, the time is synced from NTP periodly.  When this fails Tasmota functions based on time stop working.

**Reply by sfromis** [reply]
With failing Wifi, Tasmota may restart after some time, meaning that relying on assumptions of clock still being available is not a good strategy.

**Reply by pkkrusty** [reply]
You can set time manually, but the esp internal clock can drift pretty quickly. You could add a DS3231 to keep time which is very accurate. 

I'm not a fan of the 30 minute - 1 hour restart when wifi not found, and a very simple change in the code could make it a #define where that restart could be suppressed if desired.

**Reply by SimonFili** [reply]
Thanks for participating!  

I'm not sure that Tasmota restart after 30 min to 1h when no Wi-Fi.  I have an ESP8266 running rules since 16 hours ago, with Wi-Fi disconnected.  

Also, the timing of the LED (rules) still seems on time (+/- 2sec) since 16h ago with no Wi-Fi.

**Reply by pkkrusty** [reply]
On boot, if no wifi found, I think you will see a restart in the 30min-1hr window (based on device MAC address). If you have wifi then disconnect, seems like that restart doesn't happen. Test the "no wifi on boot" to confirm restart. You could even grab the MAC address and figure exactly when you can expect the restart.

**Comment by SimonFili** [comment]
Ok!  This is not good!  I wanted to be sure the timers would work without internet.  This is for a vertical garden tower. Pumps for water and lights.  I will have to revisit the design, I guess!

**Reply by sfromis** [reply]
You could add a RTC or GPS chip as a real time source, after being synchronized to start. Does not cost much.

**Reply by SimonFili** [reply]
I'm using a Sonoff DUAL R3, so this is not an easy option.

**Reply by SimonFili** [reply]
For this specific purpose, I would maybe attempt to program the ESP32 directly in Arduino to ensure I can run those timers without internet.  Yes the time would drift, without RTC, but I could implement a Wi-Fi fixed SSID hot spot procedure from the user point of view to have the device connected to it and re-sync the NTP.  But obviously, an RTC would be best!

That type of board https://www.aliexpress.com/item/1005005292859647.html?spm=a2g0o.store_pc_home.hotSpots_2006121659530.1
Would be perfect for this, then.

**Reply by Jason2866** [reply]
The maker of this board uses Tasmota a lot. He is now and than in Tasmota Discord to ask what stuff (sensors etc...) is supported from Tasmota. So his design choices are using most of the time Tasmota supported hardware. Overall his hardware quality is really good.

**Comment by SimonFili** [comment]
Alright, so I configured an ESP8266 to connect to my phone hotspot.
Connected the USB and watching the "Seriallog 4"

Then made a rule to open a relay every minute, same syntax as above with a pulsetime 20

I'm checking the logs now and the ESP is trying to connect in a loop to the Wi-Fi hotspot, but can't because I turned it off
**The rule works, the LED (relay) is flashing every minute for 2 seconds.**

So, this is on an ESP8266, my other one was on ESP32, so is there a difference in firmware that can explain that?

**Reply by SimonFili** [reply]
Ok something interesting!  The rule was still working after 20 minutes or so.
So I pressed the reset button on the nodemcu and the retry connection to Wi-Fi started right away, but the rule did not run.
I then open the hotspot, the ESP connected after a while and the rule just started back.

So, the fact that the rule does not run, is not simply because the Wi-Fi is off.

I then turned off the hotspot again, without a reset.  The rule is running.

**Reply by sfromis** [reply]
It is indeed not immediate stop of operations when wifi fails, but as I wrote *after some time*.

**Reply by SimonFili** [reply]
Still going after 3h with no Wi-Fi.  I will leave it until tomorrow AM.

**Reply by SimonFili** [reply]
Ok the rule is still running after 13 hours.  So "after some time" can be "forever" ;-)

I'm really not trying to be annoying here, but it seems to me that if the device is booting/reset and no Wi-Fi is connected, then the timers/rules do not work.

If timers/rules were running and Wi-Fi connected, then Wi-Fi disconnect, then rules will continue "for a while", which seems to be a long time.

I will leave this setup running until tonight, which will be around 24h mark.  I expect the rules to be running ok.

**Reply by pkkrusty** [reply]
I think it's a time issue. On a restart, utctime() is never set, so rules that need a time trigger won't work. If you add a boot rule to set an arbitrary time (`on system#init do time 167134338 endon`), the rules will probably start working. Note that system#boot won't work because that won't trigger until time is set via NTP. 

If you can ensure the power NEVER goes out, you can initialize your unit with wifi (and thus NTP). But that's not a very robust solution.

**Reply by SimonFili** [reply]
Wow!  Thanks for this, make senses.

2 questions:
1) If I have an RTC chip, I assume the time will be set correctly and rules would be OK, even if no Wi-Fi/NTP is available. Right? 
2) You said on the other thread, grab the MAC to figure out the restart time.  How do you calculate this from the MAC?

Thanks again!

**Reply by pkkrusty** [reply]
In support_wifi.ino, the following lines manage the auto restart:

if (100 == Wifi.max_retry) {  // Restart after 100 * (WIFI_RETRY_OFFSET_SEC + MAC) / 2 seconds
            TasmotaGlobal.restart_flag = 2;
}

and Wifi.retry_init = WIFI_RETRY_OFFSET_SEC + (ESP_getChipId() & 0xF);

So the offset_secs (20) plus the last digit of the MAC address (ranging from 0-15), so interval is 20-35 times 100 divided by 2 = 1000 - 1750 seconds (16.6 minutes to 29.2 minutes)

1. Correct.

**Reply by SimonFili** [reply]
Alright, thanks for the details.  Like I said previously and tested, if I restart (reset button) the ESP, while not connected to Wi-Fi, the rules do not start.  So this make sense with this statement "On a restart, utctime() is never set"

**On the other hand**, my test ESP, ran the rules OK for more than 16 hours without Wi-Fi.  So, if they're supposed to have an auto-restart after a maximum of 30 minutes, then the rules would have stopped after 30 minutes, but it did not.

Am I missing something?

**Reply by pkkrusty** [reply]
the test ESP had no access to wifi from the moment of startup? Or started with wifi and then you turned it off?

I believe the auto-restart happens on startup if the ESP doesn't find the wifi network, it will only search for a limited time before restarting and trying again (because rarely, a underlying esp-idf wifi stack error occurs that Tasmota can't deal with, see #18888)

If it had network access on startup, a flag is set, and the max_retry count-up doesn't happen, thus no restart.

**Reply by SimonFili** [reply]
The only network configured on the ESP is my phone hot-spot.  It's off, so "ESP had no access to wifi from the moment of startup"

Ok this "If it had network access on startup, a flag is set, and the max_retry count-up doesn't happen, thus no restart." 
Explain the rest.

Thanks

**Reply by pkkrusty** [reply]
Best bet would be to connect via serial and monitor what is happening, that way you can be sure. If there was no wifi from startup, the clock could not have been set, so time#minute rules should never have triggered.

Hard to troubleshoot someone else's device without logs.

**Reply by SimonFili** [reply]
Yes I did that, just did not post hours of logs ;-)
Just checked, the rules are still running with no Wi-Fi, after 2 days.  
The timing seems to be off by less than 2 seconds

**Reply by tavi747** [reply]
Hi,

A simple countdown timer should not need internet connection. I wanted to use a rule, on multiple Tasmota flashed plugs, to power up my home lab in a certain sequence. Ofc, when power down there is no Wi-Fi. So I guess, time based rules are pretty useless.
